Service Fabric integration docs (#577)
microsoft/reverse-proxy: docs/docfx/articles/service-fabric-int.md

## Key YARP integration features
- Reverse proxy supporting HTTP/2, gRPC and WebSockets
- Advanced routing in SF cluster
- A variety of load balancing algorithms

## Update the project file
Open the Project and find the `ItemGroup` referencing the YARP package, then add a reference to YARP.ServiceFabric package next to it.
 
 ```XML
<ItemGroup> 
  <PackageReference Include="Microsoft.ReverseProxy" Version="1.0.0-preview.8.*" />
  <PackageReference Include="Microsoft.ReverseProxy.ServiceFabric" Version="1.0.0-preview.8.*" />
</ItemGroup> 
```

## Update Startup
SF integration is plugged into the rest of YARP via a special `IProxyConfigProvider` and various support service. All of that can be registered in the Asp.Net Core DI container with the following code:
```C#
public void ConfigureServices(IServiceCollection services)
{
    services.AddControllers();
    services.AddReverseProxy()
        .LoadFromServiceFabric();

    services.Configure<ServiceFabricDiscoveryOptions>(_configuration.GetSection("ServiceFabricDiscovery"));

> }

Author: Done.

### Config example
The following is an example of an `appsettings.json` file with `ServiceFabricDiscovery` section.
```JSON
{
  "ServiceFabricDiscovery": {
    "DiscoveryPeriod": "00:00:05",
    "DiscoverInsecureHttpDestinations": true,
    "AlwaysSendImmediateHealthReports": true
  }
}
```
It can be loaded into memory on the proxy startup with the following code method called from within Asp.Net's `ConfigureServices` method:
```C#
services.Configure<ServiceFabricDiscoveryOptions>(_configuration.GetSection("ServiceFabricDiscovery"));
```

## SF service enlistment
YARP integration is enabled and configured per each SF service. The configuration is specified in the `Service Manifest` as a service extension named `YARP-preview` containing a set of labels defining specific YARP parameters. The parameter's name is set as `Key` attribute of `<Label>` element and the value is the element's content.

These are the supported parameters:
- `YARP.Enable` - indicates whether the service opt-ins to serving traffic through YARP. Default `false`
- `YARP.EnableDynamicOverrides` - indicates whether application parameters replacement is enabled on the service. Default `false`

> - `YARP.Routes.<routeName>.Path` - configures path-based route matching. The value directly assigned to [ProxyMatch.Path](xref:Microsoft.ReverseProxy.Abstractions.ProxyMatch) property and the standard route matching logic will be applied.

Author: Added a separate 'Route definitions' section.

> - `YARP.Routes.<routeName>.Host` - configures `Host` header based route matching. Multiple hosts should be separated by comma `,`. The value is split into a list of host names which is then directly assigned to [ProxyMatch.Hosts](xref:Microsoft.ReverseProxy.Abstractions.ProxyMatch) property and the standard route matching logic will be applied.

Author: Submitted #588.

> - `YARP.Routes.<routeName>.Host` - configures `Host` header based route matching. Multiple hosts should be separated by comma `,`. The value is split into a list of host names which is then directly assigned to [ProxyMatch.Hosts](xref:Microsoft.ReverseProxy.Abstractions.ProxyMatch) property and the standard route matching logic will be applied.

Author: Added a separate 'Route definitions' section.

- `YARP.Backend.HealthCheck.Active.*` - configures YARP active health checks to be run against the given service. Available parameters and their meanings are provided on [the respective documentation page](xref:dests-health-checks.md). Optional parameter
- `YARP.Backend.HealthCheck.Passive.*` - configures YARP passive health checks to be run against the given service. Available parameters and their meanings are provided on [the respective documentation page](xref:dests-health-checks.md). Optional parameter
- `YARP.Backend.Metadata.*` - sets the cluster's metadata. Optional parameter
- `YARP.Backend.BackendId` - overrides the cluster's Id. Default cluster's Id is the SF service name. Optional parameter
- `YARP.Backend.ServiceFabric.ListenerName` - sets an explicit listener name controlling selection of the main service's endpoint for each replica/instance that is used to route client requests to and is stored on the `Destination.Address` property in YARP's model. Optional parameter

> - `YARP.Backend.HealthCheck.Active.ServiceFabric.ListenerName` - sets an explicit listener name controlling selection of the health probing endpoint for each replica/instance that is used to probe replica/instance health state and is stored on the `Destination.Health` property in YARP's model. Optional parameter

Author: Moved `ListenerName` close to `Active.*` parameter explanation. Clarified why it's a special case. Submitted #590.

- `YARP.Backend.ServiceFabric.StatefulReplicaSelectionMode` - sets statefull replica selection mode. Supported values `All`, `PrimaryOnly`, `SecondaryOnly`. Values `All` and `SecondaryOnly` mean that the active secondary replicas will also be eligible for getting all kinds of client requests including writes. Default value `All`

> NOTE: Label values can use the special syntax `[AppParamName]` to reference an application parameter with the name given within square brackets. This is consistent with Service Fabric conventions, see e.g. [using parameters in Service Fabric](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-how-to-specify-port-number-using-parameters).

### Service extension example
```diff
 <ServiceManifest Name="Service1Pkg" Version="1.0.0" xmlns="http://schemas.microsoft.com/2011/01/fabric">
   <ServiceTypes>
     <StatelessServiceType ServiceTypeName="Service1Type" >
       <Extensions>
+        <Extension Name="YARP-preview">

> +          <Labels xmlns="http://schemas.microsoft.com/2015/03/fabact-no-schema">

Author: It seems no. But, as far as I see it using schemaless extensions appears to be common. In example, [Traefik's integration extension](https://doc.traefik.io/traefik/v1.7/configuration/backends/servicefabric/) is also schemaless.

> +            <Label Key="YARP.Enable">true</Label>

Author: Submitted #591.

+            <Label Key="YARP.Routes.route1.Path">{**catchall}</Label>
+            <Label Key='YARP.Backend.HealthCheck.Active.Enabled'>true</Label>
+            <Label Key='YARP.Backend.HealthCheck.Active.Timeout'>30</Label>

> +            <Label Key='YARP.Backend.HealthCheck.Active.Interval'>10</Label>

Reviewer: We might need specific doc entries for some of these properties that explain the expected value format. Are these in seconds? We used TimeSpans in config (00:00:00).

Author: Submitted #589.

+            <Label Key='YARP.Backend.HealthCheck.Active.Policy'>ConsecutiveFailures</Label>
+          </Labels>
+        </Extension>
       </Extensions>
     </StatelessServiceType>
   </ServiceTypes>
 
   <!-- ... -->
 </ServiceManifest>
```

## Dynamic YARP model

> YARP.ServiceFabric dynamically constructs and updates YARP runtime model based on the SF topology metadata. On start, it connects to SF cluster, enumerates the deployed entities (e.g applications, services, etc.), maps to YARP configuration model, validates and applies it. After the initial configuration has been applied, it starts a backround tasks periodically polling SF cluster for topology changes. The polling frequency can be set in the YARP configuration. The Service Fabric to YARP entities mapping is shown in the following table.

Reviewer: ```suggestion
YARP.ServiceFabric dynamically constructs and updates the YARP runtime model based on the SF topology metadata. On start, it connects to the SF cluster it's running in, enumerates the deployed entities (e.g applications, services, etc.), maps to the YARP configuration model, validates and applies it. After the initial configuration has been applied, YARP starts a background tasks periodically polling the SF cluster for topology changes. The polling frequency can be set in the YARP configuration. The Service Fabric to YARP entities mapping is shown in the following table.
```

Service Fabric | YARP 
:--- | :---
Cluster | n/a
Application | n/a
Service Type | n/a
Named Service Instance | Cluster (ClusterId=ServiceName)
Partition | *TBD later*
Replica / Instance (one endpoint only) | Destination (Address=instance' or replica's endpoint)
YARP.Routes.<routeName>.* in ServiceManifest | ProxyRoute (id=ServiceName+routeName, Match=Hosts,Path extracted from the labels)

## Known limitations
Limitations of the current Service Fabric to YARP configuration model conversion implementation:

> - Partitioning is not supported. Partitions are enumerated to retrieve all nested replicas/instances, but partioning key is not handled in any way. Specifically, depending of how YARP routing is configured, it's possible to route a request having one partion key (e.g 'A') to a replica of another partition (e.g. 'B').

Reviewer: ```suggestion
- Partitioning is not supported. Partitions are enumerated to retrieve all of the nested replicas/instances, but the partitioning key is not handled in any way. Specifically, depending on how YARP routing is configured, it's possible to route a request having one partition key (e.g 'A') to a replica of another partition (e.g. 'B').
```

> - Only one endpoint per each SF service's replica/instance is considered and gets converted to a [Destination](xref:Microsoft.ReverseProxy.Abstractions.Destination).

Reviewer: Which endpoint is picked by default if there are multiple? Is that what the `YARP.Backend.ServiceFabric.ListenerName` parameter is for? And the same name is used for all replicas?

Author: Submitted #590.

- All statefull service replica roles are treated equally. No special differentiation logic is applied. Depending on the configuration, active secondary replicas can also be converted to `Destinations` and directly receive client requests.

> - Each of named service instance get converted to separate YARP's [Clusters](xref:Microsoft.ReverseProxy.Abstractions.Cluster) which are completely unrelated to each other.

Reviewer: ```suggestion
- Each of the named service instances get converted to separate YARP [Clusters](xref:Microsoft.ReverseProxy.Abstractions.Cluster) which are completely unrelated to each other.
```

> - Naive error handling. It ignores most of the failures.

Reviewer: ```suggestion
- Limited error handling. Most failures are logged and suppressed. Some errors will prevent the proxy config from being updated.
```
Is that accurate?

Author: Yes.

# Service Fabric Integration

Introduced: preview8

YARP can be integrated with Service Fabric as a reverse proxy managing HTTP/HTTPS traffic ingress to a Service Fabric cluster, including support for gRPC and Web Sockets. Currently, the integration module is shipped as a separate package and has a limited support of SF availability and scalability scenarios, but it will be gradually improved over time to support more advanced SF deployment schemes.

## Key YARP integration features
- Reverse proxy supporting HTTP/2, gRPC and WebSockets
- Advanced routing in SF cluster
- A variety of load balancing algorithms

## Update the project file

> Open the Project and find the `ItemGroup` referencing the YARP package, then add a reference to YARP.ServiceFabric package next to it.

Reviewer: ```suggestion
Open the Project and find the `ItemGroup` referencing the YARP package, then add a reference to the ServiceFabric package next to it.
```

# Service Fabric Integration
YARP can be integrated with Service Fabric as a reverse proxy managing HTTP/HTTPS traffic ingress to a Service Fabric cluster, including support for gRPC and Web Sockets. Currently, the integration module is shipped as a separate package and has a limited support of SF availability and scalability scenarios, but it will be gradually improved over time to support more advanced SF deployment schemes.

## Key YARP integration features
- Reverse proxy supporting HTTP/2, gRPC and WebSockets
- Advanced routing in SF cluster

> - Sophisticated load balancing algorithms

Reviewer: I don't think that any of our [LoadBalancingModes](https://github.com/microsoft/reverse-proxy/blob/master/src/ReverseProxy/Abstractions/ClusterDiscovery/Contract/LoadBalancingMode.cs) qualify as sophisticated 😁.

Author: They are kind of comparing to the SF built-in ones. :) Though, if you want to have the wording changed, I'm open to suggestions.

Reviewer: "A variety of load balancing algorithms"?

Author: Fixed.

> 

Reviewer: And sections here that show adding the package dependency and startup code, like the sections we have in the getting started guide:
https://microsoft.github.io/reverse-proxy/articles/getting_started.html#update-the-project-file

Author: Added 2 sections.

> ## Integration component configuration

Reviewer: Include an example JSON config?

Reviewer: Call out that these are all available on ServiceFabricDiscoveryOptions as well, and show how to call that.
https://github.com/microsoft/reverse-proxy/blob/c81cf0bc0a65dd5c79f4b7015f7fb8a679d9adb4/samples/ReverseProxy.ServiceFabric.Sample/Startup.cs#L35

- `DiscoveryPeriod` - SF cluster topology metadata polling period. Default `00:00:30`
- `AllowStartBeforeDiscovery` - indicates whether YARP can start before the initial SF topology to YARP runtime model conversion completes. Setting it to `true` can lead to having the proxy started with an inconsistent configuration. Default `false`
- `DiscoverInsecureHttpDestinations` - indicates whether to discover unencrypted HTTP (i.e. non-HTTPS) endpoints. Default `false`
- `AlwaysSendImmediateHealthReports` - indicates whether SF service health reports are always sent immediately or only when explicitly requested. This flag controls only 'healthy' reports behavior. 'Unhealthy' ones are always sent immediately. Default `false`

## SF service enlistment
YARP integration is enabled and configured per each SF service. The configuration is specified in the `Service Manifest` as a service extension named `YARP-preview` containing a set of labels defining specific YARP parameters. The parameter's name is set as `Key` attribute of `<Label>` element and the value is the element's content.

> There are the following supported parameters:

Reviewer: ```suggestion
These are the supported parameters:
```

Reviewer: Which parameters are required? Either Host or Path?

Can you specify multiple routes for one service?

Author: I explained how to list multiple hosts and added the comments that Path and Hosts are set to the respective `ProxyMatch.Path` and `ProxyMatch.Hosts` properties and then our standard routing logic applies.

- `YARP.Enable` - indicates whether the service opt-ins to serving traffic through YARP. Default `false`
- `YARP.Routes.<routeName>.Path` - configures path-based route matching
- `YARP.Routes.<routeName>.Host` - configures `Host` header based route matching
- `YARP.Backend.Healthcheck.Active.*` - configures YARP active health checks to be run against the given service. Available parameters and their meanings are provided on [the respective documentation page](xref:dests-health-checks.md)

> > NOTE: Label values can use the special syntax `[AppParamName]` to reference an application parameter with the name given within square brackets. This is consistent with Service Fabric conventions, see e.g. [using parameters in Service Fabric](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-how-to-specify-port-number-using-parameters).

Reviewer: Do we need an example of this useage in this doc?

Author: Not sure, it's a SF-specific mechanism and the text above explicitly says that app parameter references are used in label values. Seems clear enough for SF users.

+            <Label Key='YARP.Backend.Healthcheck.Active.Policy'>ConsecutiveFailures</Label>
+          </Labels>
+        </Extension>
       </Extensions>
     </StatelessServiceType>
   </ServiceTypes>
 
   <!-- ... -->
 </ServiceManifest>
```

## Dynamic YARP model
YARP.ServiceFabric dynamically constructs and updates YARP runtime model based on the SF topology metadata. On start, it connects to SF cluster, enumerates the deployed entities (e.g applications, services, etc.), maps to YARP configuration model, validates and applies it. After the initial configuration has been applied, it starts a backround tasks periodically polling SF cluster for topology changes. The polling frequency can be set in the YARP configuration. The Service Fabric to YARP entities mapping is shown in the following table.

Service Fabric | YARP 
:--- | :---
Cluster | n/a
Application | n/a
Service Type | n/a
Named Service Instance | Cluster (ClusterId=ServiceName)
Partition | *TBD later*
Replica / Instance | Destination (Address=instance' or replica's endpoint)
YARP.Routes.<routeName>.* in ServiceManifest | ProxyRoute (id=ServiceName+routeName, Match=Hosts,Path extracted from the labels)

> 

Reviewer: Do you have any advice to share or link to about testing SF locally?

Author: There is a guideline about creating a sample SF service project in [ServiceFabric.Sample readme](https://github.com/microsoft/reverse-proxy/blob/master/samples/ReverseProxy.ServiceFabric.Sample/README.md). It looks sufficient. Or do you have any specific suggestions?

Reviewer: The repo samples aren't published as part of our docs, they're mainly for our benefit.

It would help to have a small section that says a YARP SF project can be tested locally by installing the SF tools from [here](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-get-started) and setting up a service using the above example manifest.

Author: Added 'Testing SF integration locally' section.
